### DISCUSSION TITLE
Allow disabling server source maps locally

### DISCUSSION THREAD

**Discussion by alex-statsig** [vercel/next.js#62198] [Ideas]
### Goals

Reduce high memory overhead in development when server modules are repeatedly recompiled (for no real reason, see https://github.com/vercel/next.js/issues/45204) and held in memory

### Background

I've observed during memory profiling that many duplicated copies of source maps appear in memory profiler. In my case, the memory held for a single API route is 60MB. This is duplicated multiple times as the API route is recompiled in development, per API route, leading to an eventual OOM not long after.

<img width="1176" alt="Screenshot 2024-02-17 at 11 10 24 AM" src="https://github.com/vercel/next.js/assets/77301670/3a00af6e-1567-4c34-8730-e3184c90f944">

```
""use strict";
/*
 * ATTENTION: An "eval-source-map" devtool has been used.
 * This devtool is neither made for production nor for readable output files.
 * It uses "eval()" calls to create a separate source file with attached SourceMaps in the browser devtools.
 * If you are trying to read the output file, select a different devtool (https://webpack.js.org/configuration/devtool/)
 * or disable the default devtool with "devtool: false".
 * If you are looking for production-ready output files, see mode: "production" (https://webpack.js.org/configuration/mode/).
 */
(() => {
var exports = {};
exports.id = "pages/api/graphql";
exports.ids = ["pages/api/graphql"];
exports.modules = {
...
```

### Proposal

There could be a next.config option to disable the maps.

Additionally I've tried manually setting webpack.devtool and got "Warning: Reverting webpack devtool to 'eval-source-map'. Changing the webpack devtool in development mode will cause severe performance regressions.".

**Comment by j23schoen** [comment]
hey @alex-statsig have you found any ways to fix this? I'm experiencing the same thing locally. My heap snapshot has up to 200mb in just the `(string)` entry and I haven't found a way to fix it

**Reply by alex-statsig** [reply]
I have not overall. I was able to track down the source of our duplication at least which helped a lot. We were adding some listeners / setting some intervals on module imports / otherwise closure-leaking the module in really hard-to-spot ways. Next.js doesn't provide great ways to manage this cleanup (I think because the assumption is a serverless backend setup), but was able to track these down with some memory profiling dominator trees and ensure they got cleaned up or ran in a way that didnt leak scope.
